src/net_processing.h

```diff
@@ -77,8 +78,14 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
     /** Whether this node ignores txs received over p2p. */
     virtual bool IgnoresIncomingTxs() = 0;
 
-    /** Relay transaction to all peers. */
-    virtual void RelayTransaction(const uint256& txid, const uint256& wtxid) = 0;
+    /** Schedule a transaction to be relayed to all peers at a later time. */
+    virtual void ScheduleTxForRelayToAll(const uint256& txid, const uint256& wtxid) = 0;
+
+    /**
+     * Schedule a local transaction to be relayed. This is done asynchronously
+     * either via short-lived privacy connections or via `ScheduleTxForRelayToAll()`.
+     */
+    virtual void ScheduleLocalTxForRelay(const uint256& txid, const uint256& wtxid) = 0;
```

Comment:
_moving discussion from https://github.com/bitcoin/bitcoin/pull/27509#issuecomment-1659008138_

> 3. Restarting the node and reloading the wallet logged messages about pushing the unbroadcast TXs to the mempool, but I do not think any sensitive-relay connections were triggered to deliver those. Eventually I tried sending another new TX and then, even though only 5 sensitive connections were being opened for that one new TX, over time those connections were used for some of the backlog ("pretending peer requested tx..."). Until the 5 sensitive connections dropped down to 0. **update: eventually `ReattemptInitialBroadcast` triggered and the backlog increased the count to 100 sensitive connections

This looks to be the same in `master` (I only read the code). Maybe it can be improved. When a new transaction arrives, `BroadcastTransaction()` is called which adds it to the mempool (via `node.mempool->AddUnbroadcastTx()`) and also schedules it for relay (via `node.peerman->ScheduleLocalTxForRelay()`).

However after a restart, `LoadMempool()` only adds it to the mempool (via `pool.AddUnbroadcastTx()`). `BroadcastTransaction()` is not called and neither `ScheduleLocalTxForRelay()` is called.

This should improve it:

<details>
<summary>[patch] relay all unbroadcast transactions at startup</summary>

```diff
diff --git i/src/net_processing.cpp w/src/net_processing.cpp
index b541cba0b9..1300d0daf0 100644
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -510,13 +510,14 @@ public:
     bool ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt) override
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, !m_recent_confirmed_transactions_mutex, !m_most_recent_block_mutex, !m_headers_presync_mutex, g_msgproc_mutex);
     bool SendMessages(CNode* pto) override
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex, !m_recent_confirmed_transactions_mutex, !m_most_recent_block_mutex, g_msgproc_mutex);
 
     /** Implement PeerManager */
-    void StartScheduledTasks(CScheduler& scheduler) override;
+    void StartScheduledTasks(CScheduler& scheduler) override
+        EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
     void CheckForStaleTipAndEvictPeers() override;
     std::optional<std::string> FetchBlock(NodeId peer_id, const CBlockIndex& block_index) override
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
     bool GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) const override EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
     bool IgnoresIncomingTxs() override { return m_opts.ignore_incoming_txs; }
     void SendPings() override EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
@@ -1869,22 +1870,33 @@ PeerManagerImpl::PeerManagerImpl(CConnman& connman, AddrMan& addrman,
         m_txreconciliation = std::make_unique<TxReconciliationTracker>(TXRECONCILIATION_VERSION);
     }
 }
 
 void PeerManagerImpl::StartScheduledTasks(CScheduler& scheduler)
 {
+    AssertLockNotHeld(m_peer_mutex);
+
     // Stale tip checking and peer eviction are on two different timers, but we
     // don't want them to get out of sync due to drift in the scheduler, so we
     // combine them in one function and schedule at the quicker (peer-eviction)
     // timer.
     static_assert(EXTRA_PEER_CHECK_INTERVAL < STALE_CHECK_INTERVAL, "peer eviction timer should be less than stale tip check timer");
     scheduler.scheduleEvery([this] { this->CheckForStaleTipAndEvictPeers(); }, std::chrono::seconds{EXTRA_PEER_CHECK_INTERVAL});
 
     // schedule next run for 10-15 minutes in the future
     const std::chrono::milliseconds delta = 10min + GetRandMillis(5min);
     scheduler.scheduleFromNow([&] { ReattemptInitialBroadcast(scheduler); }, delta);
+
+    if (m_opts.sensitiverelayowntx) {
+        for (const auto& txid : m_mempool.GetUnbroadcastTxs()) {
+            CTransactionRef tx = m_mempool.get(txid);
+            if (tx != nullptr) {
+                ScheduleLocalTxForRelay(txid, tx->GetWitnessHash());
+            }
+        }
+    }
 }
 
 /**
  * Evict orphan txn pool entries based on a newly connected
  * block, remember the recently confirmed transactions, and delete tracked
  * announcements for them. Also save the time of the last tip update and
```
</details>

I did not include it in this PR yet. What do you think?

It looks like an improvement in `master` too, but I only did it if sensitive relay is used because the purpose of this PR is not to change the behavior when `-sensitiverelayowntx=0` compared to `master`.

Reply:
OK I see. In that case, it can be a follow-up. Probably there is some reason we keep the scheduler and don't resend right on startup (we wouldn't want spy nodes detecting our restart! ...?)